### PR TITLE
deploy hook for Ruckus ZoneDirector / Unleashed

### PR DESCRIPTION
**Pull request by kchiem**
Tested on my Unleashed install, but according to the code it's adapted form, it should work for ZD installs too.

### CODE REVIEW

**Comment by uberjay** [comment]
Just stumbled across this and verified it works for deploying certificate updates from OPNsense to my Ruckus Unleashed controller.

Steps:

1) Place `ruckus.sh` file in `/usr/local/share/examples/acme.sh/deploy/`
2) `chmod 555 ruckus.sh`
3) Set up certificate for ruckus (via the OPNsense GUI). let's call the domain name `$domain`.
4) Run issue/renew certificate action.
5) ssh to your OPNsense system as root and drop to a shell
6) find the corresponding certificate configuration directory under `/var/etc/acme-client/certs/`
    * in my case it was `/var/etc/acme-client/certs/6716ac20201125.16873601`.
    * let's call this full directory path `$certdir`.
7) edit the cert config file (this will be `$certdir/$domain/$domain.conf`) to add the three variables:
    * `RUCKUS_HOST`
    * `RUCKUS_USER`
    * `RUCKUS_PASS`
8) execute the acme.sh deploy action:

```
/usr/local/sbin/acme.sh --deploy --syslog 7 --debug --server 'letsencrypt' \
  --home '/var/etc/acme-client/home' \
  --cert-home "$certdir/$domain" \
  --certpath "$certdir/$domain/cert.pem" \
  --keypath "$certdir/$domain/private.key" \
  --capath "$certdir/$domain/chain.pem" \
  --fullchainpath "$certdir/$domain/fullchain.pem" \
  --domain "$domain" --deploy-hook ruckus
```

So, I vote for including this deployment hook. What else is needed to make it happen? It needs to be upstream (here) before OPNsense will add an option to use it, and it'd be really fantastic to have it "just work". ✨

File: deploy/ruckus.sh [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  _debug RUCKUS_USER "$RUCKUS_USER"
+  _debug RUCKUS_PASS "$RUCKUS_PASS"
+
+  COOKIE_JAR=$(mktemp)
+  cleanup() {
+    rm $COOKIE_JAR
+  }
+  trap cleanup EXIT
+
+  LOGIN_URL=$(curl https://$RUCKUS_HOST -ksSLo /dev/null -w '%{url_effective}')
```
Comment (Neilpang) [inline deploy/ruckus.sh:75]: do not use `curl`,  please use `_post()` or `_get()` function instead.

**Comment by fraenki** [comment]
@uberjay I don't know if the original author is still working on this... Maybe you would be interested to finish the job? I'd suggest to apply the proposed fixes and submit a new PR.

**Comment by ms264556** [comment]
> @uberjay I don't know if the original author is still working on this... Maybe you would be interested to finish the job? I'd suggest to apply the proposed fixes and submit a new PR.

I'm the author of the [code this is adapted from](https://ms264556.net/ruckus/PfSenseLetsEncryptToRuckus).

If anyone's already working on a fix then let me know. Otherwise I'm happy to re-implement this correctly: I'm on holiday for a couple more months, so I have the time.

(I had assumed this PR wouldn't be accepted as-is, since it's not idiomatic with the other acme code).

**Comment by kchiem** [comment]
@ms264556 I was planning on looking into it now, but if you're able to, please do go ahead.

**Comment by uberjay** [comment]
@ms264556 That'd be awesome! I don't know if the original PR author is following along here, and I don't personally have the bandwidth. It'd be very much appreciated!

**Comment by ms264556** [comment]
> @ms264556 I was planning on looking into it now, but if you're able to, please do go ahead.

Sure. I'll setup some test APs and have a look this afternoon.

I do remember that the acme code for existing integrations was very low level compared to my lazy 'get curl do all the thinking' code, so I don't think this is a 5 minute job.

File: deploy/ruckus.sh (hunk 1 above)
Comment (ms264556) [reply deploy/ruckus.sh]: My reimplemention using _get() and post() has been pushed to this PR now

File: deploy/ruckus.sh [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    return 1
+  fi
+
+  _savedeployconf RUCKUS_HOST "$RUCKUS_HOST"
+  _savedeployconf RUCKUS_USER "$RUCKUS_USER"
+  _savedeployconf RUCKUS_PASS "$RUCKUS_PASS"
+
+  _debug RUCKUS_HOST "$RUCKUS_HOST"
+  _debug RUCKUS_USER "$RUCKUS_USER"
+  _debug RUCKUS_PASS "$RUCKUS_PASS"
```
Comment (Neilpang) [inline deploy/ruckus.sh:66]: use `_secure_debug` for password, so that you won't leak it.

File: deploy/ruckus.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+  export HTTPS_INSECURE=1
```
Comment (Neilpang) [inline deploy/ruckus.sh:68]: why do we need "insecure" here?

File: deploy/ruckus.sh [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  _post_upload "uploadcert" "$_cfullchain"
+  
+  _info "Uploading private key"
+  _post_upload "uploadprivatekey" "$_ckey"
+
+  _info "Replacing certificate"
+  _replace_cert_ajax='<ajax-request action="docmd" comp="system" updater="rid.0.5" xcmd="replace-cert" checkAbility="6" timeout="-1"><xcmd cmd="replace-cert" cn="'$RUCKUS_HOST'"/></ajax-request>'
+  _post "$_replace_cert_ajax" "$_base_url/_cmdstat.jsp" >/dev/null
+  
+  info "Rebooting"
```
Comment (Neilpang) [inline deploy/ruckus.sh:130]: use `_info`

File: deploy/ruckus.sh (hunk 3 above)
Comment (ms264556) [reply deploy/ruckus.sh:68]: Ruckus devices ship with 1024 bit self-signed server certificates, so curl will fail with error 60 unless I set HTTPS_INSECURE=1.

File: deploy/ruckus.sh (hunk 2 above)
Comment (ms264556) [reply deploy/ruckus.sh]: Fixed!

File: deploy/ruckus.sh (hunk 4 above)
Comment (ms264556) [reply deploy/ruckus.sh]: Fixed!

**Comment by ms264556** [comment]
Oops. Would help if I quoted all of my _info() args. One more commit incoming

File: deploy/ruckus.sh (hunk 2 above)
Comment (ms264556) [reply deploy/ruckus.sh]: Fixed

File: deploy/ruckus.sh (hunk 4 above)
Comment (ms264556) [reply deploy/ruckus.sh]: Fixed, and I also fixed the unquoted _info strings

File: deploy/ruckus.sh (hunk 3 above)
Comment (uberjay) [reply deploy/ruckus.sh:68]: Overriding HTTPS_INSECURE to always be enabled seems wrong to me -- for initial deployment one can add `--insecure`: `acme.sh –deploy -d your.domain.here –deploy-hook ruckus –insecure`. On subsequent cert deploy runs, HTTPS_INSECURE=1 should no longer be necessary.

**Comment by kchiem** [comment]
@ms264556 https://github.com/acmesh-official/acme.sh/actions/runs/11811051778/job/32909254460

File: deploy/ruckus.sh (hunk 3 above)
Comment (ms264556) [reply deploy/ruckus.sh]: No problem. That's fixed now.

**Comment by ms264556** [comment]
> @ms264556 [acmesh-official/acme.sh/actions/runs/11811051778/job/32909254460](https://github.com/acmesh-official/acme.sh/actions/runs/11811051778/job/32909254460)

Fixed

**Comment by Neilpang** [comment]
please add the usage here:
https://github.com/acmesh-official/acme.sh/wiki/deployhooks

**Comment by ms264556** [comment]
> please add the usage here: [Wiki: deployhooks](https://github.com/acmesh-official/acme.sh/wiki/deployhooks)

That's done

**Comment by ms264556** [comment]
I noticed in the dev docs that we're not supposed to use awk. Our awk usage isn't doing anything useful, so I've given @kchiem a PR to remove this, which should be pushed here soon

**Comment by uberjay** [comment]
@ms264556 Thanks for taking care of this. Is there anything else that's needed, apart from acme.sh project approval/merge?

I'm excited at the prospect of this being bundled with OPNsense.

**Pull request merged**